File: tests/test_telegram_direct.js
// Tests for lib/telegram_direct.js (ADR-085).
// Uses Node's built-in test runner (node:test), same convention as
// tests/test_n8n_notify.js. No real network calls — global.fetch is
// mocked and restored around every test.
//
//   node --test tests/test_telegram_direct.js

const test = require('node:test');
const assert = require('node:assert/strict');

const {
  sendTelegramMessage, buildProductReadyMessage, buildSaleMadeMessage, buildCriticalErrorMessage,
} = require('../lib/telegram_direct.js');

function withMockedFetch(impl, fn) {
  const original = global.fetch;
  global.fetch = impl;
  return fn().finally(() => { global.fetch = original; });
}

function withEnv(vars, fn) {
  const original = {};
  for (const k of Object.keys(vars)) { original[k] = process.env[k]; process.env[k] = vars[k]; }
  return fn().finally(() => {
    for (const k of Object.keys(vars)) {
      if (original[k] === undefined) delete process.env[k];
      else process.env[k] = original[k];
    }
  });
}

test('missing token/chatId -> no-ops, never calls fetch, reports error not throw', async () => {
  let fetchCalled = false;
  await withMockedFetch(
    async () => { fetchCalled = true; throw new Error('fetch must not be called'); },
    async () => {
      const result = await sendTelegramMessage('hello', { token: null, chatId: null });
      assert.equal(result.sent, false);
      assert.match(result.error, /not configured/);
      assert.equal(fetchCalled, false);
    }
  );
});

test('real send success -> reports message_id, posts real chat_id/text payload', async () => {
  await withMockedFetch(
    async (url, opts) => {
      assert.equal(url, 'https://api.telegram.org/bottest-token/sendMessage');
      const body = JSON.parse(opts.body);
      assert.equal(body.chat_id, '999');
      assert.equal(body.text, 'hello world');
      return { ok: true, json: async () => ({ ok: true, result: { message_id: 55 } }) };
    },
    async () => {
      const result = await sendTelegramMessage('hello world', { token: 'test-token', chatId: '999' });
      assert.deepEqual(result, { sent: true, message_id: 55, error: null });
    }
  );
});

test('telegram rejection -> reported, not thrown', async () => {
  await withMockedFetch(
    async () => ({ ok: false, status: 400, json: async () => ({ ok: false, description: 'chat not found' }) }),
    async () => {
      const result = await sendTelegramMessage('hello', { token: 't', chatId: 'c' });
      assert.equal(result.sent, false);
      assert.equal(result.error, 'chat not found');
    }
  );
});

test('network failure -> reported, not thrown', async () => {
  await withMockedFetch(
    async () => { throw new Error('network down'); },
    async () => {
      const result = await sendTelegramMessage('hello', { token: 't', chatId: 'c' });
      assert.equal(result.sent, false);
      assert.equal(result.error, 'network down');
    }
  );
});

test('falls back to process.env when token/chatId not passed explicitly', async () => {
  await withEnv({ TELEGRAM_BOT_TOKEN: 'env-token', OPENCLAW_TELEGRAM_CHAT_ID: 'env-chat' }, async () => {
    await withMockedFetch(
      async (url) => {
        assert.equal(url, 'https://api.telegram.org/botenv-token/sendMessage');
        return { ok: true, json: async () => ({ ok: true, result: { message_id: 1 } }) };
      },
      async () => {
        const result = await sendTelegramMessage('hi');
        assert.equal(result.sent, true);
      }
    );
  });
});

test('buildProductReadyMessage includes niche and price when present', () => {
  const msg = buildProductReadyMessage({
    niche: 'AI compliance automation', production_id: 'PROD-123',
    pricing_strategy: { recommended_price: 388 },
  });
  assert.match(msg, /منتج جاهز للمراجعة/);
  assert.match(msg, /AI compliance automation/);
  assert.match(msg, /\$388/);
  assert.match(msg, /PROD-123/);
});

test('buildProductReadyMessage omits price line cleanly when absent, never shows undefined', () => {
  const msg = buildProductReadyMessage({ niche: 'x', production_id: 'PROD-1', pricing_strategy: null });
  assert.doesNotMatch(msg, /undefined/);
  assert.doesNotMatch(msg, /السعر/);
});

test('buildSaleMadeMessage includes platform and amount', () => {
  const msg = buildSaleMadeMessage({ platform: 'Paddle', amount: 388, title: 'AI Compliance System' });
  assert.match(msg, /بيع جديد/);
  assert.match(msg, /Paddle/);
  assert.match(msg, /\$388/);
});

test('buildCriticalErrorMessage shows first reason and a count of extras', () => {
  const msg = buildCriticalErrorMessage(['reason one', 'reason two', 'reason three']);
  assert.match(msg, /عطل حرج/);
  assert.match(msg, /reason one/);
  assert.match(msg, /\+2/);
  assert.doesNotMatch(msg, /reason two/); // only the first is shown verbatim, never all of them
});

test('buildCriticalErrorMessage with a single reason has no "+N extra" line', () => {
  const msg = buildCriticalErrorMessage(['only reason']);
  assert.doesNotMatch(msg, /\+\d/);
});

File: tests/test_telegram_direct.py
"""Tests for channels/telegram_direct.py (ADR-085).

Runs with stdlib unittest. No live Telegram API call is ever made --
requests.post is patched in every test that would otherwise hit the
network.

    python -m unittest tests.test_telegram_direct -v
"""

import sys
import unittest
from pathlib import Path
from unittest.mock import patch, MagicMock

_FACTORY_ROOT = Path(__file__).resolve().parent.parent
if str(_FACTORY_ROOT) not in sys.path:
    sys.path.insert(0, str(_FACTORY_ROOT))

from channels import telegram_direct


def _fake_env(tmp_path, token="test-token", chat_id="12345"):
    env_file = tmp_path / ".env"
    lines = []
    if token is not None:
        lines.append(f"TELEGRAM_BOT_TOKEN={token}")
    if chat_id is not None:
        lines.append(f"OPENCLAW_TELEGRAM_CHAT_ID={chat_id}")
    env_file.write_text("\n".join(lines), encoding="utf-8")
    return env_file


class TestSendTelegramMessage(unittest.TestCase):
    def test_missing_config_is_reported_not_raised(self):
        import tempfile
        with tempfile.TemporaryDirectory() as d:
            env_file = _fake_env(Path(d), token=None, chat_id=None)
            result = telegram_direct.send_telegram_message("hello", env_path=env_file)
        self.assertFalse(result["sent"])
        self.assertIn("not configured", result["error"])

    def test_real_send_success_reports_message_id(self):
        import tempfile
        with tempfile.TemporaryDirectory() as d:
            env_file = _fake_env(Path(d))
            fake_response = MagicMock(ok=True, content=b'{"ok": true, "result": {"message_id": 42}}')
            fake_response.json.return_value = {"ok": True, "result": {"message_id": 42}}
            with patch.object(telegram_direct.requests, "post", return_value=fake_response) as mock_post:
                result = telegram_direct.send_telegram_message("hello", env_path=env_file)
        self.assertTrue(result["sent"])
        self.assertEqual(result["message_id"], 42)
        self.assertIsNone(result["error"])
        # chat_id and text must actually reach Telegram's real payload shape
        _, kwargs = mock_post.call_args
        self.assertEqual(kwargs["json"]["chat_id"], "12345")
        self.assertEqual(kwargs["json"]["text"], "hello")

    def test_telegram_rejection_is_reported_not_raised(self):
        import tempfile
        with tempfile.TemporaryDirectory() as d:
            env_file = _fake_env(Path(d))
            fake_response = MagicMock(ok=False, status_code=400, content=b'{"ok": false, "description": "chat not found"}')
            fake_response.json.return_value = {"ok": False, "description": "chat not found"}
            with patch.object(telegram_direct.requests, "post", return_value=fake_response):
                result = telegram_direct.send_telegram_message("hello", env_path=env_file)
        self.assertFalse(result["sent"])
        self.assertEqual(result["error"], "chat not found")

    def test_network_failure_is_reported_not_raised(self):
        import tempfile
        with tempfile.TemporaryDirectory() as d:
            env_file = _fake_env(Path(d))
            with patch.object(telegram_direct.requests, "post", side_effect=telegram_direct.requests.RequestException("timeout")):
                result = telegram_direct.send_telegram_message("hello", env_path=env_file)
        self.assertFalse(result["sent"])
        self.assertIn("timeout", result["error"])


if __name__ == "__main__":
    unittest.main()
